ops/lib.rs

```diff
@@ -347,172 +341,166 @@ fn codegen_fast_impl(
+      let op_state_decl = if use_op_state_refcell {
+        quote! { let #op_state_name = ctx.state.clone(); }
       } else {
-        let output = if returns_result {
-          get_fast_result_return_type(&f.sig.output)
-        } else {
-          let output = &f.sig.output;
-          quote! { #output }
-        };
-        let func_name = format_ident!("func_{}", name);
-        let op_state_name = if use_op_state {
-          input_idents.first().unwrap().clone()
-        } else {
-          quote! { op_state }
-        };
-        let recv_decl = if use_op_state || returns_result {
-          quote! {
-            // SAFETY: V8 calling convention guarantees that the callback options pointer is non-null.
-            let opts: &mut #core::v8::fast_api::FastApiCallbackOptions = unsafe { &mut *fast_api_callback_options };
-            // SAFETY: data union is always created as the `v8::Local<v8::Value>` version.
-            let data = unsafe { opts.data.data };
-            // SAFETY: #core guarantees data is a v8 External pointing to an OpCtx for the isolates lifetime
-            let ctx = unsafe {
-              &*(#core::v8::Local::<#core::v8::External>::cast(data).value()
-              as *const #core::_ops::OpCtx)
-            };
-            let #op_state_name = &mut ctx.state.borrow_mut();
-          }
-        } else {
-          quote! {}
+        quote! { let #op_state_name = &mut ctx.state.borrow_mut(); }
+      };
+      quote! {
+        // SAFETY: V8 calling convention guarantees that the callback options pointer is non-null.
+        let opts: &mut #core::v8::fast_api::FastApiCallbackOptions = unsafe { &mut *fast_api_callback_options };
+        // SAFETY: data union is always created as the `v8::Local<v8::Value>` version.
+        let data = unsafe { opts.data.data };
+        // SAFETY: #core guarantees data is a v8 External pointing to an OpCtx for the isolates lifetime
+        let ctx = unsafe {
+          &*(#core::v8::Local::<#core::v8::External>::cast(data).value()
+          as *const #core::_ops::OpCtx)
         };
+        #op_state_decl
+      }
+    } else {
+      quote! {}
+    };
 
-        let result_handling = if returns_result {
-          quote! {
-            match result {
-              Ok(result) => {
-                result
-              },
-              Err(err) => {
-                #op_state_name.last_fast_op_error.replace(err);
-                opts.fallback = true;
-                Default::default()
-              },
+    let (trampoline, raw_block) = if is_async {
+      (
+        quote! {
+          fn #func_name (_recv: #core::v8::Local<#core::v8::Object>, __promise_id: i32, __promise_resolver: v8::Local<v8::Function>, #(#inputs),*) -> i32 {
+            #recv_decl
+            let __op_id = ctx.id;
+            let isolate = unsafe { &mut *ctx.isolate };
+            let resolver = #core::v8::Global::<#core::v8::Function>::new(isolate, __promise_resolver);
```

Comment:
Does this not assert? Holy moly, I would've 100% assumed creating globals would be forbidden by Fast API asserts.

Reply:
From V8 dev: Creating should be Global is fine (although perhaps not quite what is expected). cc @MayaLekova